commit 9dba0dbae0d81c4ed548a62136deaffd199d86a6 (refs/changes/09/2009/1)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2017-05-26T12:10:49-07:00 (2 years, 4 months ago)
    
    AGENT-1072 cn-agent crashed with "RangeError: Maximum call stack size exceeded"

diff --git a/lib/heartbeater.js b/lib/heartbeater.js
index 4fee73e..ec257d8 100644
--- a/lib/heartbeater.js
+++ b/lib/heartbeater.js
@@ -474,7 +474,14 @@ StatusReporter.prototype.gatherDiskUsage = function (vms, callback) {
                 cb(err);
             }
 
-            function onDataset(dataset, next) {
+            function onDataset(dataset, _next) {
+                // Guard against us blowing up the stack (AGENT-1072)
+                var next = function (err) {
+                    setImmediate(function (_err) {
+                        _next(_err);
+                    }, err);
+                };
+
                 // Eliminate snapshots and sub-filesystems
                 var UUID_RE = '([0-9a-f]{8}-[0-9a-f]{4}-' +
                               '[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-' +
diff --git a/package.json b/package.json
index a6c9f80..d708cf2 100644
--- a/package.json
+++ b/package.json
@@ -6,7 +6,7 @@
   "private": true,
   "dependencies": {
     "assert-plus": "1.0.0",
-    "async": "1.5.2",
+    "async": "2.4.1",
     "bunyan": "1.8.5",
     "ctype": "0.5.4",
     "digest-stream": "0.2.2",
diff --git a/tools/jsl.node.conf b/tools/jsl.node.conf
index 0b8ae95..c44e8ee 100644
--- a/tools/jsl.node.conf
+++ b/tools/jsl.node.conf
@@ -119,6 +119,7 @@
 +define process
 +define require
 +define setInterval
++define setImmediate
 +define setTimeout
 +define Buffer
 +define JSON
